src/include/daos/stack_mmap.h

```diff
@@ -0,0 +1,49 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * This file is part of daos
+ *
+ * src/include/daos/stack_mmap.h
+ */
+
+#ifdef ULT_MMAP_STACK
+#include <sys/mman.h>
+#include <abt.h>
+
+/* mmap()'ed stacks can allow for a bigger size with no impact on
+ * memory footprint if unused
+ */
+#define MMAPED_ULT_STACK_SIZE (2 * 1024 * 1024)
+
+/* ABT_key for mmap()'ed ULT stacks */
+extern ABT_key stack_key;
+
+/* since being allocated before start of stack its size must be a
+ * multiple of (void *) !!
+ */
+typedef struct {
+	void *stack;
+	size_t stack_size;
+	void (*thread_func)(void *);
+	void *thread_arg;
+	d_list_t stack_list;
+} mmap_stack_desc_t;
```

Comment:
If this is only used when the region is on the free list then why does it need thread_func and thread_arg here, or even stack for that matter, would stack_list and stack_size be sufficient?

On that note do we need to reserve space for this when the stack is in use?

Reply:
I agree with @ashleypittman , the thread_func and thread_arg shouldn't be here, and I don't quite see why it need another allocation, can't we just pass them as arguments when creating ULT?

Reply:
> I agree with @ashleypittman , the thread_func and thread_arg shouldn't be here, and I don't quite see why it need another allocation, can't we just pass them as arguments when creating ULT?

again, the wrapper mmap_stack_wrapper() can only have a single argument when we call ABT_thread_create() and this can only be the mmap_stack_desc_t, so we need to store the original+final thread_func/thread_arg there !

Reply:
Oh, I see your point now. So the side effect of this patch is that all the ULT functions will become mmap_stack_wrapper(), that'll devalue the watchdog, since watchdog will always report the same ULT function address/symbol, and we won't be able to tell what's the real ULT function anymore.

Reply:
> So the side effect of this patch is that all the ULT functions will become mmap_stack_wrapper(), that'll devalue the watchdog, since watchdog will always report the same ULT function address/symbol, and we won't be able to tell what's the real ULT function anymore.

Well, you are right, but I should be able to restore this capability by using ABT_thread_get_attr() and ABT_thread_attr_get_stack(), in place of ABT_thread_get_thread_func() within sched_watchdog_prep(), to be able to retrieve ULT's mmap_stack_desc_t and its thread_[func,arg] values .

Reply:
In fact, since I still need to use mmap_stack_wrapper() to set ABT_key, which uses mmap_stack_desc_t as ULT argument, to make watchdog code work, it is easier to get arg via ABT_thread_get_arg() and thus retrieve the original function address from mmap_stack_desc_t instead to compute its address from stack address returned by ABT_thread_get_attr()+ABT_thread_attr_get_stack() !!...